When the archive can't be found in a `bbs2gh` migration, suggest using the `--bbs-shared-home` option

In `bbs2gh` migrations, we download the generated archive from the Bitbucket Server instance using SSH or SMB.

If we look for the file and it doesn't exist, we output a message like this:

> [16:40] [ERROR] Source export archive (/var/atlassian/application-data/bitbucket/shared/data/migration/export/Bitbucket_export_1.tar) does not exist.

https://github.com/github/gh-gei/blob/a8ddc6960494f5b57a5eb86497bed477a3ebe7ac/src/bbs2gh/Services/BbsSshArchiveDownloader.cs#L94

If this happens, the most likely reason is that Bitbucket's data is not located in the default home directory, `/var/atlassian/application-data/bitbucket`. You can point to a non-default home directory by specifying the `--bbs-shared-home` directory.

We should update this error message to explain where we are looking for the file, how to check the correct `$BITBUCKET_HOME` path and how to use `--bbs-shared-home` to customise the search location.

This change may also apply to SMB, but I'm not 100% sure.


## Todo
<!--
List of todos (generally not issues) to complete this task. Please use checkboxes so that progress is visible.
-->
- [ ] ...

## Dependencies
<!--
List issues which this issue depends on.
-->
- ...

--- a/RELEASENOTES.md
+++ b/RELEASENOTES.md
@@ -1,5 +1,6 @@
 - Improve error message when `migrate-repo` is used with a target personal access token (PAT) with insufficient permissions
 - Ensure `--no-ssl-verify` flag is honored when downloading archives from GHES
 - `--bbs-project` and `--bbs-repo` are now both required in `gh bbs2gh migrate-repo` command when `--bbs-server-url` is set
-* Added `--keep-archive` flag to `gh gei migrate-repo` and `gh gei generate-script`. When migrating from GHES this will skip the step where we delete the archive from your machine, leaving it around as a local file.
+- Added `--keep-archive` flag to `gh gei migrate-repo` and `gh gei generate-script`. When migrating from GHES this will skip the step where we delete the archive from your machine, leaving it around as a local file.
 - Continue to next mannequin mapping in `gh gei reclaim-mannequin --csv` if a username doesn't exist
+- Display more helpful message when the Bitbucket export archive is not found when using `gh bbs2gh migrate-repo`

--- a/src/bbs2gh/Services/BbsSmbArchiveDownloader.cs
+++ b/src/bbs2gh/Services/BbsSmbArchiveDownloader.cs
@@ -208,10 +208,8 @@ public sealed class BbsSmbArchiveDownloader : IBbsArchiveDownloader
             : throw new OctoshiftCliException(
                 $"Couldn't create SMB file handle for \"{sharedFilePath}\" (Status Code: {status})." +
                 (IsObjectPathNotFoundStatus(status) && BbsSharedHomeDirectory is BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS
-                    ? " This usually happens because the archive could not be found at the default Bitbucket's shared home " +
-                      $"which is located in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX}\" if Bitbucket is installed on a Linux machine " +
-                      $"or in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS}\" if installed on a Windows machine. " +
-                      "You can point to a non default shared directory by specifying the --bbs-shared-home option."
+                    ? "This most likely means that your Bitbucket instance uses a non-default Bitbucket shared home directory, so we couldn't find your archive. " +
+                      "You can point the CLI to a non-default shared directory by specifying the --bbs-shared-home option."
                     : ""));
     }
 

--- a/src/bbs2gh/Services/BbsSmbArchiveDownloader.cs
+++ b/src/bbs2gh/Services/BbsSmbArchiveDownloader.cs
@@ -205,7 +205,14 @@ public sealed class BbsSmbArchiveDownloader : IBbsArchiveDownloader
 
         return IsSuccessStatus(status)
             ? sharedFileHandle
-            : throw new OctoshiftCliException($"Couldn't create SMB file handle for \"{sharedFilePath}\" (Status Code: {status}).");
+            : throw new OctoshiftCliException(
+                $"Couldn't create SMB file handle for \"{sharedFilePath}\" (Status Code: {status})." +
+                (IsObjectPathNotFoundStatus(status) && BbsSharedHomeDirectory is BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS
+                    ? " This usually happens because the archive could not be found at the default Bitbucket's shared home " +
+                      $"which is located in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX}\" if Bitbucket is installed on a Linux machine " +
+                      $"or in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS}\" if installed on a Windows machine. " +
+                      "You can point to a non default shared directory by specifying the --bbs-shared-home option."
+                    : ""));
     }
 
     private long? GetFileSize(ISMBFileStore fileStore, object sharedFileHandle)
@@ -218,4 +225,6 @@ public sealed class BbsSmbArchiveDownloader : IBbsArchiveDownloader
     private bool IsSuccessStatus(NTStatus status) => status is NTStatus.STATUS_SUCCESS;
 
     private bool IsEndOfFileStatus(NTStatus status) => status is NTStatus.STATUS_END_OF_FILE;
+
+    private bool IsObjectPathNotFoundStatus(NTStatus status) => status is NTStatus.STATUS_OBJECT_PATH_NOT_FOUND;
 }

--- a/src/bbs2gh/Services/BbsSshArchiveDownloader.cs
+++ b/src/bbs2gh/Services/BbsSshArchiveDownloader.cs
@@ -91,7 +91,14 @@ public sealed class BbsSshArchiveDownloader : IBbsArchiveDownloader, IDisposable
 
         if (!_sftpClient.Exists(sourceExportArchiveFullPath))
         {
-            throw new OctoshiftCliException($"Source export archive ({sourceExportArchiveFullPath}) does not exist.");
+            throw new OctoshiftCliException(
+                $"Source export archive ({sourceExportArchiveFullPath}) does not exist." +
+                (BbsSharedHomeDirectory is BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX
+                    ? " This usually happens because the archive could not be found at the default Bitbucket's shared home " +
+                      $"which is located in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX}\" if Bitbucket is installed on a Linux machine " +
+                      $"or in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS}\" if installed on a Windows machine. " +
+                      "You can point to a non default shared directory by specifying the --bbs-shared-home option."
+                    : ""));
         }
 
         _fileSystemProvider.CreateDirectory(targetDirectory);

--- a/src/bbs2gh/Services/BbsSshArchiveDownloader.cs
+++ b/src/bbs2gh/Services/BbsSshArchiveDownloader.cs
@@ -94,10 +94,8 @@ public sealed class BbsSshArchiveDownloader : IBbsArchiveDownloader, IDisposable
             throw new OctoshiftCliException(
                 $"Source export archive ({sourceExportArchiveFullPath}) does not exist." +
                 (BbsSharedHomeDirectory is BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX
-                    ? " This usually happens because the archive could not be found at the default Bitbucket's shared home " +
-                      $"which is located in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_LINUX}\" if Bitbucket is installed on a Linux machine " +
-                      $"or in \"{BbsSettings.DEFAULT_BBS_SHARED_HOME_DIRECTORY_WINDOWS}\" if installed on a Windows machine. " +
-                      "You can point to a non default shared directory by specifying the --bbs-shared-home option."
+                    ? "This most likely means that your Bitbucket instance uses a non-default Bitbucket shared home directory, so we couldn't find your archive. " +
+                      "You can point the CLI to a non-default shared directory by specifying the --bbs-shared-home option."
                     : ""));
         }
 

